### PR TITLE
New and improved iteration APIs for `Chunk`s

### PR DESCRIPTION
**Pull request by teh-cmc**
Just bunch of brand new and/or improved iteration APIs for `Chunk`s.

This will be used all over the place with the new query APIs, which will appear in a future PR.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6989?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6989?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/6989)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/store/re_chunk/src/iter.rs [hunk 1]
```diff
@@ -1,29 +1,179 @@
...
+
+        // NOTE: No need for validity checks here, `iter_offsets` already takes care of that.
+        Either::Right(
+            self.iter_component_offsets(component_name)
+                .map(move |(idx, len)| &values[idx..idx + len]),
+        )
+    }
+}
+
+// TODO(cmc): get rid of this
```
Comment (teh-cmc) [inline crates/store/re_chunk/src/iter.rs:169]: This happens later in that PR train -- at the moment this is still used all over the place.

File: crates/store/re_chunk/src/slice.rs [hunk 2]
```diff
@@ -5,17 +5,55 @@ use arrow2::array::{
...
-use crate::{Chunk, ChunkTimeline};
+use crate::{Chunk, ChunkTimeline, RowId};
 
 // ---
 
 // NOTE: Not worth writing tests for all of these, until some subtle bug comes up.
 // Most of them are indirectly stressed by our higher-level query tests anyhow.
 
 impl Chunk {
+    /// Returns the cell corresponding to the specified [`RowId`] for a given [`ComponentName`].
```
Comment (teh-cmc) [inline crates/store/re_chunk/src/slice.rs:20]: That's just moved code

File: crates/store/re_chunk/src/slice.rs [hunk 3]
```diff
@@ -493,3 +547,135 @@ impl ChunkTimeline {
         )
     }
 }
+
+// ---
+
+#[cfg(test)]
```
Comment (teh-cmc) [inline crates/store/re_chunk/src/slice.rs:553]: That's just moved code

File: crates/store/re_chunk/src/slice.rs [hunk 4]
```diff
@@ -307,10 +345,26 @@ impl Chunk {
             components: components
                 .iter()
                 .map(|(&component_name, list_array)| {
-                    (
-                        component_name,
-                        crate::util::filter_array(list_array, &validity_filter),
-                    )
+                    let filtered = crate::util::filter_array(list_array, &validity_filter);
```
Comment (teh-cmc) [inline crates/store/re_chunk/src/slice.rs:348]: Nice newly added lil' optimization

File: crates/store/re_chunk/src/iter.rs [hunk 5]
```diff
@@ -1,29 +1,179 @@
...
+    ) -> impl Iterator<Item = (TimeInt, RowId)> + '_ {
+        let Some(list_array) = self.components.get(component_name) else {
+            return Either::Left(std::iter::empty());
+        };
+
+        if self.is_static() {
+            let indices = izip!(std::iter::repeat(TimeInt::STATIC), self.row_ids());
+
+            if let Some(validity) = list_array.validity() {
+                Either::Right(Either::Left(Either::Left(
```
Comment (Wumpf) [inline crates/store/re_chunk/src/iter.rs:62]: ... not sure why I whined about double nested either earlier Oo. Maybe you want to introduce new `IterVariants3` and `IterVariants4` iterators which internally match things?

File: crates/store/re_chunk/src/iter.rs [hunk 6]
```diff
@@ -1,29 +1,179 @@
...
+                )))
+            } else {
+                Either::Right(Either::Right(Either::Right(indices)))
+            }
+        }
+    }
+
+    /// Returns an iterator over the offsets (`(offset, len)`) of a [`Chunk`], for a given timeline
+    /// and component.
+    pub fn iter_component_offsets(
```
Comment (Wumpf) [inline crates/store/re_chunk/src/iter.rs:91]: It took me a while to process that these offsets & lengths are about the different array lengths of components at each event. Would be nice to mention that again for noobs

File: crates/store/re_chunk/src/iter.rs [hunk 7]
```diff
@@ -1,29 +1,179 @@
...
+                        .filter_map(|(i, o)| validity.get_bit(i).then_some(o)),
+                )))
+            } else {
+                Either::Right(Either::Right(Either::Right(indices)))
+            }
+        }
+    }
+
+    /// Returns an iterator over the offsets (`(offset, len)`) of a [`Chunk`], for a given timeline
+    /// and component.
```
Comment (Wumpf) [inline crates/store/re_chunk/src/iter.rs:90]: there's no given timeline here

**Review by Wumpf — APPROVED** [review]
so many iterators * _ *
looks good tho

File: crates/store/re_chunk/src/iter.rs [hunk 8]
```diff
@@ -1,29 +1,179 @@
...
+    ///
+    /// If the chunk is static, `timeline` will be ignored.
+    ///
+    /// See also:
+    /// * [`Self::iter_component_indices`].
+    /// * [`Self::iter_indices_owned`].
+    #[inline]
+    pub fn iter_indices(&self, timeline: &Timeline) -> impl Iterator<Item = (TimeInt, RowId)> + '_ {
+        if self.is_static() {
+            Either::Right(Either::Left(izip!(
```
[the remark this answers is not in the document]
Comment (teh-cmc) [reply crates/store/re_chunk/src/iter.rs:27]: It is not arbitrary!

They _should_ all follow this model:
* The first layer is always the emptiness layer: `Left` is empty, `Right` is non-empty.
* The second layer is the temporarily layer: `Left` is static, `Right` is temporal.
* Any layers beyond that follow the same pattern: `Left` doesn't have something, while `Right` does.

File: crates/store/re_chunk/src/iter.rs (hunk 5 above)
Comment (teh-cmc) [reply crates/store/re_chunk/src/iter.rs:62]: Tbh I'd rather keep the nicely composable `Either` than introduce more ad-hoc things.

The caller never has to deal with them anyhow :shrug:

File: crates/store/re_chunk/src/iter.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (teh-cmc) [reply crates/store/re_chunk/src/iter.rs:27]: Added a comment in the code reflecting that.

**Pull request merged**